core/domain/summary_services.py

```diff
@@ -254,15 +254,17 @@ def get_learner_collection_dict_by_id(
         next_exploration_id = collection.first_exploration_id
         completed_exp_ids = []
 
-    # Here, the return type of 'to_dict' method is CollectionDict but for
-    # implementation purpose we are assigning LearnerCollectionDict which
-    # is inherited from CollectionDict. So, due to the difference in types
-    # MyPY throws an error. Thus to avoid the error, we used ignore here.
+    # Here we use MyPy ignore because the return type of 'to_dict' method
+    # is CollectionDict but for implementation purpose we are assigning
+    # LearnerCollectionDict which is inherited from CollectionDict. So,
+    # due to the difference in types MyPy throw a error. Thus to avoid
+    # the error, we used ignore here.
     collection_dict: LearnerCollectionDict = collection.to_dict()  # type: ignore[assignment]
-    # Here, expression has type List[CollectionNodeDict] but for implementation
-    # purpose we are assigning List[LearnerCollectionNodeDict]. So, due the
-    # difference in types MyPY throws an error. Thus to avoid the error, we
-    # used ignore here.
+    # Here we use MyPy ignore because expression has type List[
```

Comment:
This doesn't seem right at all, please fix the code and drop the ignore.

Reply:
As per the code comment, the expression provides a list of collectionNodeDict but the expected type here is list of LearnerCollectionNodeDict. And LearnerCollectionNodeDict is expected because below we are accessing `exploration_summary` key which is not defined in collectionNode dict. 
https://github.com/oppia/oppia/blob/68bb129c7fe773b989670a59162534b8893789bb/core/domain/summary_services.py#L296

Reply:
If we do not change the type of collection_node['nodes'] explicitly, then the following things will happen:

If we do not change the type of collection_dict['nodes'], then it's type will become `List[collectionNodeDict]`. So, here
https://github.com/oppia/oppia/blob/20af54e0574a4f99690f15f2fbb97f2a397f22f9/core/domain/summary_services.py#L278
collection_node will have type `CollectionNodeDict` and below we are defining a new key `exploration_summary` https://github.com/oppia/oppia/blob/20af54e0574a4f99690f15f2fbb97f2a397f22f9/core/domain/summary_services.py#L296
https://github.com/oppia/oppia/blob/20af54e0574a4f99690f15f2fbb97f2a397f22f9/core/domain/summary_services.py#L298

And CollectionNodeDict does not contain this `exploration_summary` key. So, if we don not change the type of collection_dict['nodes'] explicitly to the List[LearnerCollectionNodeDict] ( this LearnerCollectionNodeDict contains exploration_summary key ) then we have to use the two ignores here.

The main thing -- is that I'm changing the type of collection_dict['nodes'] explicitly to make this dict consistent with rest of the function rather than using ignore every time we found a different key that was not defined in CollectonNodeDict.